src/services/utilities.ts

```diff
@@ -1256,8 +1256,10 @@ namespace ts {
      * Finds the rightmost token satisfying `token.end <= position`,
      * excluding `JsxText` tokens containing only whitespace.
      */
-    export function findPrecedingToken(position: number, sourceFile: SourceFile, startNode?: Node, excludeJsdoc?: boolean): Node | undefined {
-        const result = find(startNode || sourceFile);
+    export function findPrecedingToken(position: number, sourceFile: SourceFileLike, startNode: Node, excludeJsdoc?: boolean): Node | undefined;
+    export function findPrecedingToken(position: number, sourceFile: SourceFile, startNode?: Node, excludeJsdoc?: boolean): Node | undefined;

```

Comment:
Out of curiosity, why couldn't this just change to `SourceFileLike`? Seems like it'd be a backwards compatible change to accept a wider set of types.

Reply:
Oh, I didn't notice that `startNode` becomes required.

Reply:
`SourceFileLike` has no parse tree structure; it’s just a text container. So if you provide a `SourceFileLike` you _must_ provide an actual Node to start in.

Reply:
Yeah, I just didn't notice the one character `?` difference, oops.